packages/ics721/src/execute.rs

```diff
@@ -0,0 +1,473 @@
+            vec![]
+        } else {
+            let message = SubMsg::<T>::reply_on_success(
+                WasmMsg::Instantiate {
+                    admin: None,
+                    code_id: CW721_CODE_ID.load(deps.storage)?,
+                    msg: self.init_msg(&env, &class)?,
+                    funds: vec![],
+                    // Attempting to fit the class ID in the label field
+                    // can make this field too long which causes data
+                    // errors in the SDK.
+                    label: "ics-721 debt-voucher cw-721".to_string(),
+                },
+                INSTANTIATE_CW721_REPLY_ID,
+            );
+            vec![message]
+        };
+
+        // Store mapping from classID to classURI. Notably, we don't check
+        // if this has already been set. If a new NFT belonging to a class
+        // ID we have already seen comes in with new metadata, we assume
+        // that the metadata has been updated on the source chain and
+        // update it for the class ID locally as well.
+        CLASS_ID_TO_CLASS.save(deps.storage, class.id.clone(), &class)?;
+
+        let mint = WasmMsg::Execute {
+            contract_addr: env.contract.address.into_string(),
+            msg: to_binary(&ExecuteMsg::Callback(CallbackMsg::Mint {
+                class_id: class.id,
+                receiver,
+                tokens,
+            }))?,
+            funds: vec![],
+        };
+
+        Ok(Response::<T>::default()
+            .add_attribute("method", "callback_create_vouchers")
+            .add_submessages(instantiate)
+            .add_message(mint))
+    }
+
+    /// Default implementation using `cw721_base::msg::InstantiateMsg`
+    fn init_msg(&self, env: &Env, class: &Class) -> StdResult<Binary> {
+        to_binary(&cw721_base::msg::InstantiateMsg {
+            // Name of the collection MUST be class_id as this is how
+            // we create a map entry on reply.
+            name: class.id.clone().into(),
+            symbol: class.id.clone().into(),
+            minter: env.contract.address.to_string(),
+        })
+    }
+
+    /// Helper function in case creator is provided in `ClassData`. In this case, creator comes from another chain
+    /// and needs to be converted to the target chain.
+    fn get_creator(&self, env: &Env, class: &Class) -> StdResult<Option<String>> {
+        match class.data.clone() {
+            // in case no class data is provided, ics721 will be used as the creator
+            None => Ok(None),
+            Some(data) => {
+                // class data may be of type `ClassData` or some other custom class data.
+                let class_data_result: StdResult<ClassData> = from_binary(&data);
+                if class_data_result.is_err() {
+                    return Ok(None);
+                }
+                let class_data = class_data_result.unwrap();
+
+                match class_data.owner {
```

Comment:
Can you put this matching code block into a named function something like `convert_owner_chain_address`. Im not opinionated about the name.

Reply:
Done. See in utils.rs